component/loki/source/file/file.go

```diff
@@ -0,0 +1,302 @@
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", "loki.source.file component shutting down, stopping readers")
+		for _, r := range c.readers {
+			r.Stop()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			for _, receiver := range c.receivers {
+				receiver <- entry
+			}
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.args = newArgs
+	c.receivers = newArgs.ForwardTo
+
+	oldPaths := make(map[string]struct{})
+
+	// Stop all readers and recreate them below. This avoids the issue we saw
+	// with stranded wrapped handlers staying behind until they were GC'ed and
+	// sending duplicate message to the global handler. It also makes sure that
+	// we update everything with the new labels. Simply zeroing out the
+	// c.readers map did not work correctly to shut down the wrapped handlers
+	// in time.
+	// TODO (@tpaschalis) We should be able to optimize this somehow and eg.
+	// cache readers for paths we already know about, and whose labels have not
+	// changed. Once we do that we should:
+	// a) Call to c.pruneStoppedReaders to give cached but errored readers a
+	// chance to restart.
+	// b) Stop tailing any files that were no longer in the new targets
+	// and conditionally remove their readers only by calling toStopTailing
+	// and c.stopTailingAndRemovePosition.
+	for p, r := range c.readers {
+		oldPaths[p] = struct{}{}
+		r.Stop()
+	}
+	c.readers = make(map[string]reader)
+
+	if len(newArgs.Targets) == 0 {
+		level.Debug(c.opts.Logger).Log("msg", "no files targets were passed, nothing will be tailed")
+		return nil
+	}
+
+	for _, target := range newArgs.Targets {
+		path := target[pathLabel]
+		c.reportSize(path)
+
+		var labels = make(model.LabelSet)
+		for k, v := range target {
+			if strings.HasPrefix(k, model.ReservedLabelPrefix) {
+				continue
+			}
+			labels[model.LabelName(k)] = model.LabelValue(v)
+		}
+
+		handler := loki.AddLabelsMiddleware(labels).Wrap(loki.NewEntryHandler(c.handler, func() {}))
+
+		reader, err := c.startTailing(path, handler)
+		if err != nil {
+			continue
+		}
+
+		c.readers[path] = reader
+	}
+
+	// Remove from the positions file any paths that had a Reader before, but
+	// are no longer in the updated set of Targets.
+	for path := range missing(c.readers, oldPaths) {
+		c.posFile.Remove(path)
+	}
+
+	return nil
+}
+
+// DebugInfo returns information about the status of tailed targets.
+func (c *Component) DebugInfo() interface{} {
+	var res readerDebugInfo
+	for _, reader := range c.readers {
+		path := reader.Path()
+		offset, _ := c.posFile.Get(path)
+		res.TargetsInfo = append(res.TargetsInfo, targetInfo{
+			Path:       path,
+			IsRunning:  reader.IsRunning(),
+			ReadOffset: offset,
+		})
```

Comment:
Can this include the full set of labels from the target?

Reply:
Also, do we have access to the last time a log line was read to expose it here?

Reply:
> The component does not exist yet (but will exist by the next release). Should I remove the example and add it back when it's ready?

It'd be much easier if/when we key readers both by path and labels; for now, we'd have to loop through the targets and match the path, which has a caveat.  
For targets with the same path, when creating the reader the last one wins. Here the first one wins, so it might be misleading.

> Also, do we have access to the last time a log line was read to expose it here?
No we don't, but I can add it.

Reply:
👍 Returning to decorate this with more information later sounds good to me.

Reply:
Ok, will add a TODO for now.